modules/gapi/test/own/rmat_tests.cpp

```diff
@@ -0,0 +1,143 @@
+        }
+    }
+    virtual cv::GMatDesc desc() const override { return cv::descr_of(m_mat); }
+};
+
+class RMatAdapterCopy : public RMat::Adapter {
+    cv::Mat& m_deviceMat;
+    cv::Mat  m_hostMat;
+    bool& m_callbackCalled;
+
+public:
+    RMatAdapterCopy(cv::Mat& m, bool&callbackCalled)
+        : m_deviceMat(m), m_hostMat(m.clone()), m_callbackCalled(callbackCalled)
+    {}
+    virtual cv::MatView access(RMat::Access access) const override {
+        if (access == RMat::Access::W) {
+            return MatView(cv::descr_of(m_hostMat), m_hostMat.data, m_hostMat.step,
+                           [this](){
+                               EXPECT_FALSE(m_callbackCalled);
+                               m_callbackCalled = true;
+                               m_hostMat.copyTo(m_deviceMat);
+                           });
+        } else {
+            m_deviceMat.copyTo(m_hostMat);
+            return MatView(cv::descr_of(m_hostMat), m_hostMat.data, m_hostMat.step);
+        }
+    }
+    virtual cv::GMatDesc desc() const override { return cv::descr_of(m_hostMat); }
+};
+
+void randomizeMat(cv::Mat& m) { cv::randu(m, cv::Scalar::all(127), cv::Scalar::all(40)); }
+
+template <typename RMatAdapterT>
+struct RMatTest {
+    RMatTest()
+        : deviceMat(8,8,CV_8UC1)
+        , rmat(make_rmat<RMatAdapterT>(deviceMat, callbackCalled)) {
+        randomizeMat(deviceMat);
+        expectNoCallbackCalled();
+    }
+
+    cv::Mat deviceMat;
+    bool callbackCalled = false;
+    cv::RMat rmat;
+
+    void expectCallbackCalled() { EXPECT_TRUE(callbackCalled); }
+    void expectNoCallbackCalled() { EXPECT_FALSE(callbackCalled); }
+
+    void expectDeviceDataEqual(const cv::Mat& mat) {
+        EXPECT_EQ(0, cv::norm(mat, deviceMat, NORM_INF));
+    }
+    void expectDeviceDataNotEqual(const cv::Mat& mat) {
+        EXPECT_NE(0, cv::norm(mat, deviceMat, NORM_INF));
+    }
+};
+} // anonymous namespace
+
+template<typename T>
+struct RMatTypedTest : public ::testing::Test, public T {};
+
+using RMatTestTypes = ::testing::Types< RMatTest<RMatAdapterRef>
+                                      , RMatTest<RMatAdapterCopy>
+                                      >;
+
+TYPED_TEST_CASE(RMatTypedTest, RMatTestTypes);
+
+TYPED_TEST(RMatTypedTest, Smoke) {
+    auto view = this->rmat.access(RMat::Access::R);
+    auto matFromDevice = cv::Mat(view.size(), view.type(), view.ptr());
+    EXPECT_TRUE(cv::descr_of(this->deviceMat) == this->rmat.desc());
+    this->expectDeviceDataEqual(matFromDevice);
+}
+
+static Mat wrapViewByMat(MatView& view) {
+    return Mat(view.size(), view.type(), view.ptr(), view.step());
+}
+
+TYPED_TEST(RMatTypedTest, BasicWorkflow) {
+    {
+        auto view = this->rmat.access(RMat::Access::R);
+        this->expectDeviceDataEqual(wrapViewByMat(view));
+    }
+    this->expectNoCallbackCalled();
+
+    cv::Mat dataToWrite = this->deviceMat.clone();
+    randomizeMat(dataToWrite);
+    this->expectDeviceDataNotEqual(dataToWrite);
```

Comment:
It seems it's not the best idea to expect that two randomized mats will differ at all times

Reply:
Reworked `randomizeMat`